File: 03 March/03-31/wake_up_alarm.py
import datetime as dt

def alarm_check(alarm_time, wake_time):
    SNOOZE_MINUTES = 10

    dummy_date = dt.date(1, 1, 1)

    alarm_datetime_object = dt.datetime.combine(dummy_date, dt.time.fromisoformat(alarm_time))
    wake_datetime_object = dt.datetime.combine(dummy_date, dt.time.fromisoformat(wake_time))
    snooze_datetime_object = alarm_datetime_object + dt.timedelta(minutes = SNOOZE_MINUTES)

    return "early" if wake_datetime_object < alarm_datetime_object \
        else "on time" if wake_datetime_object <= snooze_datetime_object \
        else "late"

# --- TEST SUITE ---

tests_text = r'''
1. alarm_check("07:00", "06:45") should return "early".
2. alarm_check("06:30", "06:30") should return "on time".
3. alarm_check("08:10", "08:15") should return "on time".
4. alarm_check("09:30", "09:45") should return "late".
5. alarm_check("08:15", "08:25") should return "on time".
6. alarm_check("05:45", "05:56") should return "late".
7. alarm_check("04:30", "04:00") should return "early".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-31/wakeUpAlarm.js
function alarmCheck(alarmTime, wakeTime) {
  const SNOOZE_WINDOW = "PT10M";
  
  const alarmTimeObject = Temporal.PlainTime.from(alarmTime);
  const wakeTimeObject = Temporal.PlainTime.from(wakeTime);
  const snoozeTimeObject = alarmTimeObject.add(SNOOZE_WINDOW);

  return Temporal.PlainTime.compare(wakeTimeObject, alarmTimeObject) === -1 ? "early"
    : Temporal.PlainTime.compare(wakeTimeObject, snoozeTimeObject) <= 0 ? "on time"
    : "late";
}

// --- TEST SUITE ---

const testsText = String.raw`
1. alarmCheck("07:00", "06:45") should return "early".
2. alarmCheck("06:30", "06:30") should return "on time".
3. alarmCheck("08:10", "08:15") should return "on time".
4. alarmCheck("09:30", "09:45") should return "late".
5. alarmCheck("08:15", "08:25") should return "on time".
6. alarmCheck("05:45", "05:56") should return "late".
7. alarmCheck("04:30", "04:00") should return "early".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);